rubrics/impl/src/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java

```diff
@@ -0,0 +1,797 @@
+
+        Optional<ToolItemRubricAssociation> association = Optional.empty();
+
+        Optional<Resource<ToolItemRubricAssociation>> associationResource = getRubricAssociationResource(toolId, associatedToolItemId);
+        if (associationResource.isPresent()) {
+            association = Optional.of(associationResource.get().getContent());
+        }
+        return association;
+    }
+
+
+    /**
+     * Returns the ToolItemRubricAssociation resource for the given tool and associated item ID, wrapped as an Optional.
+     * @param toolId the tool id, something like "sakai.assignment"
+     * @param associatedToolItemId the id of the associated element within the tool
+     * @return
+     */
+    protected Optional<Resource<ToolItemRubricAssociation>> getRubricAssociationResource(String toolId, String associatedToolItemId) throws Exception {
+
+        TypeReferences.ResourcesType<Resource<ToolItemRubricAssociation>> resourceParameterizedTypeReference =
+                new TypeReferences.ResourcesType<Resource<ToolItemRubricAssociation>>() {};
+
+        URI apiBaseUrl = new URI(serverConfigurationService.getServerUrl() + RBCS_SERVICE_URL_PREFIX);
+        Traverson traverson = new Traverson(apiBaseUrl, MediaTypes.HAL_JSON);
+
+        Traverson.TraversalBuilder builder = traverson.follow("rubric-associations", "search",
+                "by-tool-item-ids");
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Authorization", String.format("Bearer %s", generateJsonWebToken(toolId)));
+        builder.withHeaders(headers);
+
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("toolId", toolId);
+        parameters.put("itemId", associatedToolItemId);
+
+        Resources<Resource<ToolItemRubricAssociation>> associationResources = builder.withTemplateParameters(
+                parameters).toObject(resourceParameterizedTypeReference);
+
+        // Should only be one matching this search criterion
+        if (associationResources.getContent().size() > 1) {
+            throw new IllegalStateException(String.format(
+                    "Number of rubric association resources greater than one for request: %s",
+                    associationResources.getLink(Link.REL_SELF).toString()));
+        }
+
+        Optional<Resource<ToolItemRubricAssociation>> associationResource = associationResources.getContent().stream().findFirst();
+
+        return associationResource;
+    }
+
+    //TODO generate a public String postRubricAssociation(String tool, String id, HashMap<String,String> params)
+
+    /**
+     * Posts the rubric association.
+     * @param json The json to post.
+     * @return
+     */
+    private String postRubricResource(String targetUri, String json, String toolId) throws IOException {
+        log.debug(String.format("Post to URI '%s' body:", targetUri, json));
+
+        try {
```

Comment:
Hi, Earl. We did have these low level connections identified as clean-up items to refactor to also use Traverson (or RestTemplate directly) for consistency and deal with the resource and connection clean-up. Diego will either do that, or clean this up with a finally block as an interim step.